WordPress/Classes/Services/MediaService.m

```diff
@@ -208,6 +208,19 @@ - (void)uploadMedia:(Media *)media
     // See https://github.com/wordpress-mobile/WordPress-iOS/issues/4663
     remoteMedia.file = [remoteMedia.file stringByReplacingOccurrencesOfString:@".jpeg" withString:@".jpg"];
     NSManagedObjectID *mediaObjectID = media.objectID;
+    
+    // To avoid deadlocks we're creating a derived context instead of using self.managedObjectContext directly.
+    NSManagedObjectContext *synchronousContext = [[NSManagedObjectContext alloc] initWithConcurrencyType:NSPrivateQueueConcurrencyType];
+    synchronousContext.parentContext = self.managedObjectContext;
+    
+    [synchronousContext performBlockAndWait:^{
+        Media *mediaInContext = (Media *)[synchronousContext existingObjectWithID:mediaObjectID error:nil];
+        if (mediaInContext) {
+            mediaInContext.remoteStatus = MediaRemoteStatusPushing;
+            mediaInContext.error = nil;
+            [[ContextManager sharedInstance] saveContextAndWait:synchronousContext];
```

Comment:
I've been discussing about our CoreData Stack with @astralbodies and we feel some of the problems that are surfacing may be because our CD Stack could use a refresh.

In this particular instance, the actual problem I'm seeing is that locking a child context to synchronously save the parent contexts is a sure-recipe for deadlocks.

But from the architectura PoV: I don't think a child context should be the one to decide when a parent context is saved.  This is less of a problem in our async methods for obvious reasons.

I realize what I'm saying isn't in-line with our existing architecture, but this is one case where I feel we need to take a moment to consider addressing the core issue.

I'll think about it a bit more and post my conclusions here.

Reply:
For now I'm following the existing architecture, and letting a synchronous save operation in a child context force the parent contexts to save (asynchronously).  We'll probably review all this at the GM with @astralbodies.